source/common/quic/capsule_protocol_handler.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include "envoy/http/codec.h"
+
+#include "source/common/common/logger.h"
+#include "source/common/http/header_map_impl.h"
+
+#include "absl/strings/string_view.h"
+#include "quiche/common/simple_buffer_allocator.h"
+#include "quiche/quic/core/http/quic_spdy_stream.h"
+
+namespace Envoy {
+namespace Quic {
+
+class CapsuleProtocolHandler : public quic::QuicSpdyStream::Http3DatagramVisitor,
+                               public quiche::CapsuleParser::Visitor,
+                               protected Logger::Loggable<Logger::Id::quic_stream> {
+public:
+  // Does not take ownership of the QuicSpdyStream. "stream" must refer to a valid object
+  // that outlives CapsuleProtocolHandler.
+  explicit CapsuleProtocolHandler(quic::QuicSpdyStream* stream) : stream_(stream) {}
+
+  // quic::QuicSpdyStream::Http3DatagramVisitor
+  // Normalizes Http3 Datagrams to Capsules for Envoy to do subsequent processing.
+  void OnHttp3Datagram(quic::QuicStreamId stream_id, absl::string_view payload) override;
+
+  // quiche::CapsuleParser::Visitor
+  // Converts a given Capsule to a Http3 Datagrams and send it.
+  bool OnCapsule(const quiche::Capsule& capsule) override;
+  void OnCapsuleParseFailure(absl::string_view error_message) override;
+
+  // Re-encodes a given Capsule data into an Http3 Datagram. Returns true if the encoding is
+  // successful, otherwise false.
+  bool encodeCapsule(absl::string_view capsule_data);
+  void setStreamDecoder(Http::StreamDecoder* decoder) { stream_decoder_ = decoder; }
+  // Checks if the Capsule-Protocol header field is present in the headers to enable the
+  // Capsule-Protocol.
+  void onHeaders(Http::HeaderMapImpl* const headers);
```

Comment:
Ah, is it because Capsules are not necessary when both downstream and upstream are using H/3? Related to this question, Capsule-Protocol header value is checked only on the H/3 streams in this commit. Should we check it across the H1/H2/H3 codecs?

Reply:
Discussed this offline. I am planning to add CONNECT-UDP related logic in a separate pull request adding full CONNECT-UDP support if you don't mind.